platforms/core-runtime/messaging/src/test/groovy/org/gradle/internal/remote/internal/inet/TcpConnectorTest.groovy

```diff
@@ -321,5 +322,28 @@ class TcpConnectorTest extends ConcurrentSpec {
         acceptor?.stop()
         socketChannel.close()
     }
+    
+    @Issue("https://github.com/gradle/gradle/issues/27801")
+    def "refuses connections that don't provide a handshake"() {
+        given:
+        def connected = false
+        def action = { connected = true  }
+        def socketChannel = SocketChannel.open()
+        def acceptor = incomingConnector.accept(action, false)
+        def communicationAddress = addressFactory.getLocalBindingAddress()
+        def bindAnyPort = new InetSocketAddress(communicationAddress, 0)
+        def connectAddress = new InetSocketAddress(communicationAddress, acceptor.address.port)
+
+        when:
+        socketChannel.socket().bind(bindAnyPort)
+        socketChannel.socket().connect(connectAddress)
+        acceptor?.stop()
```

Comment:
❓ 

Null-safety is not required here, right?

Reply:
I just copied this test from the one above, which also used the null-safe access. No deeper reason :)

And yes, this is part of the scenario, because stopping the acceptor will wait for the outstanding connection requests to be handled. We want to make sure that it is handled before we assert that it was refused.